File: Data Structures & Algorithms/count-number-of-islands/submission-3.java
class Solution {
    public int numIslands(char[][] grid) {
        int count = 0;
        for (int r=0; r<grid.length; r++){
            for (int c=0; c<grid[0].length; c++){
                if (grid[r][c]=='1') {
                        count += dfs(r,c,grid);
                }
            }
        }
        return count;
        
    }

    public int dfs(int r, int c, char[][] grid) {
        if (!isValid(r, c, grid) || grid[r][c] != '1') {
            return 0;
        }

        grid[r][c] = '0';

        int[][] dirs = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

        for (int i = 0; i < dirs.length; i++) {
            int newr = r + dirs[i][0];
            int newc = c + dirs[i][1];
            if (isValid(r, c, grid) || grid[r][c] == '1') {
            dfs(newr, newc, grid);
        }
        }
        return 1;
    }

    public boolean isValid(int r, int c, char[][] grid) {
        return r >= 0 && r < grid.length && c >= 0 && c < grid[0].length;
    }

}

File: Data Structures & Algorithms/path-with-minimum-effort/submission-3.py

class Solution:
    def minimumEffortPath(self, heights: List[List[int]]) -> int:
        ROWS, COLS = len(heights), len(heights[0])
        minHeap = [(0,0,0)]
        visit = set()
        dirs = [(-1,0),(1,0),(0,1),(0,-1)]
        while minHeap:
            diff,r,c = heapq.heappop(minHeap)
            if (r,c) in visit:
                continue
            visit.add((r,c))
            if (r,c) == (ROWS-1,COLS-1):
                return diff
            for di in dirs:
                newR,newC = r+di[0],c+di[1]
                if (
                    newR < 0 or newC < 0 or
                    newR >= ROWS or newC >= COLS or
                    (newR, newC) in visit
                ):
                    continue
                newDiff = max(diff, abs(heights[r][c]-heights[newR][newC]))
                heapq.heappush(minHeap,(newDiff,newR,newC))
        return 0

